### DISCUSSION TITLE
How to install pip dependencies for a python program invoked with pyo3 from Rust

### DISCUSSION THREAD

**Discussion by ghost** [PyO3/pyo3#2400] [Questions]
Hi,

I have a basic python library which I'm calling from rust using pyo3 and this program has external dependencies which will be needed when running it from within the rust binary. After scanning the documentation I don't see a clear and clean way to do it. I see recommendations for maturin but I want to minimize the setup and reduce my crate binary size too. Any examples or guides will be greatly appreciated.

Cheers!

**Comment by l-monninger** [comment]
@nyanchor I'm wondering about roughly the same thing right now. [This](https://github.com/PyO3/pyo3/discussions/2206) makes me think the simplest way to go will likely be to just install your library as a package in your venv, and then do something like [this](https://pyo3.rs/latest/python_from_rust.html#want-to-access-python-apis-then-use-pymoduleimport).

**Reply by ghost** [reply]
@l-monninger that's interesting. I might do this. It'll just need some extra scripts to setup these dependencies at runtime for the binary to run with

**Reply by davidhewitt** [reply]
@l-monninger is correct, that's about the right approach. If you want to ship just a single binary you can pack everything up using PyOxidizer.

**Reply by Krahos** [reply]
> @l-monninger is correct, that's about the right approach. If you want to ship just a single binary you can pack everything up using PyOxidizer.

Is there any example on how to call a PyOxidizer binary? I'm trying to import a python class and its dependencies and, especially for the user, I don't want to deal with installing stuff with pip and managing virtual environments (I'm a python noob myself).

**Reply by davidhewitt** [reply]
The point as I understand it is that with PyOxidizer your users don't have to install anything at all with pip - all the packages get included directly inside the binary.

**Reply by Krahos** [reply]
Thanks, so I'll try to use their documentation or ask there on how to do this.

**Reply by pdavid31** [reply]
@Krahos I am trying to achieve the same thing. Did you manage to do so? Do you have any recommended examples, etc.?

**Reply by Krahos** [reply]
@pdavid31 Sadly no, I ended up opening an HTTP API on top of the code I had to call and moved on with my life. It wasn't ideal, but being just a prototype it was fine enough.
